File: static/js/e-commerce.js
// E-commerce functionality for Women Entrepreneurs Hub
import { initializeApp } from 'https://www.gstatic.com/firebasejs/11.0.2/firebase-app.js';
import { getAuth } from 'https://www.gstatic.com/firebasejs/11.0.2/firebase-auth.js';
import { getFirestore, collection, addDoc, updateDoc, deleteDoc, doc, getDoc, getDocs, query, where, orderBy } from 'https://www.gstatic.com/firebasejs/11.0.2/firebase-firestore.js';
import { getStorage, ref, uploadBytes, getDownloadURL } from 'https://www.gstatic.com/firebasejs/11.0.2/firebase-storage.js';

// Firebase configuration will be injected from the server
let firebaseConfig = {};
try {
  firebaseConfig = {
    apiKey: document.getElementById('firebase-config').getAttribute('data-api-key'),
    projectId: document.getElementById('firebase-config').getAttribute('data-project-id'),
    appId: document.getElementById('firebase-config').getAttribute('data-app-id'),
    authDomain: `${document.getElementById('firebase-config').getAttribute('data-project-id')}.firebaseapp.com`,
    storageBucket: `${document.getElementById('firebase-config').getAttribute('data-project-id')}.appspot.com`,
  };
} catch (error) {
  console.error('Error loading Firebase config:', error);
}

// Initialize Firebase
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
const db = getFirestore(app);
const storage = getStorage(app);

// Create a cart in localStorage if it doesn't exist
if (!localStorage.getItem('cart')) {
  localStorage.setItem('cart', JSON.stringify([]));
}

// Functions for product management
const productService = {
  // Get all products
  getAllProducts: async () => {
    try {
      const productsRef = collection(db, 'products');
      const q = query(productsRef, where('is_available', '==', true), orderBy('created_at', 'desc'));
      const querySnapshot = await getDocs(q);
      const products = [];
      
      querySnapshot.forEach((doc) => {
        products.push({
          id: doc.id,
          ...doc.data()
        });
      });
      
      return products;
    } catch (error) {
      console.error('Error getting products:', error);
      return [];
    }
  },
  
  // Get products by user ID
  getProductsByUser: async (userId) => {
    try {
      const productsRef = collection(db, 'products');
      const q = query(productsRef, where('user_id', '==', userId));
      const querySnapshot = await getDocs(q);
      const products = [];
      
      querySnapshot.forEach((doc) => {
        products.push({
          id: doc.id,
          ...doc.data()
        });
      });
      
      return products;
    } catch (error) {
      console.error('Error getting user products:', error);
      return [];
    }
  },
  
  // Get product by ID
  getProduct: async (productId) => {
    try {
      const productRef = doc(db, 'products', productId);
      const productSnap = await getDoc(productRef);
      
      if (productSnap.exists()) {
        return {
          id: productSnap.id,
          ...productSnap.data()
        };
      } else {
        console.error('Product not found');
        return null;
      }
    } catch (error) {
      console.error('Error getting product:', error);
      return null;
    }
  },
  
  // Add a new product
  addProduct: async (productData, imageFile) => {
    try {
      if (!auth.currentUser) {
        throw new Error('User must be authenticated to add products');
      }
      
      // Upload image if provided
      let imageUrl = null;
      if (imageFile) {
        const storageRef = ref(storage, `product_images/${auth.currentUser.uid}/${Date.now()}_${imageFile.name}`);
        await uploadBytes(storageRef, imageFile);
        imageUrl = await getDownloadURL(storageRef);
      }
      
      // Add product to Firestore
      const newProduct = {
        ...productData,
        user_id: auth.currentUser.uid,
        image_url: imageUrl,
        created_at: new Date(),
        updated_at: new Date(),
        is_available: true
      };
      
      const docRef = await addDoc(collection(db, 'products'), newProduct);
      
      // Send to backend to sync
      await fetch('/api/products', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          ...newProduct,
          firebase_id: docRef.id
        }),
      });
      
      return docRef.id;
    } catch (error) {
      console.error('Error adding product:', error);
      throw error;
    }
  },
  
  // Update a product
  updateProduct: async (productId, productData, imageFile) => {
    try {
      if (!auth.currentUser) {
        throw new Error('User must be authenticated to update products');
      }
      
      // Upload image if provided
      let imageUrl = productData.image_url;
      if (imageFile) {
        const storageRef = ref(storage, `product_images/${auth.currentUser.uid}/${Date.now()}_${imageFile.name}`);
        await uploadBytes(storageRef, imageFile);
        imageUrl = await getDownloadURL(storageRef);
      }
      
      // Update product in Firestore
      const productRef = doc(db, 'products', productId);
      
      const updatedProduct = {
        ...productData,
        image_url: imageUrl,
        updated_at: new Date()
      };
      
      await updateDoc(productRef, updatedProduct);
      
      // Send to backend to sync
      await fetch(`/api/products/${productId}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(updatedProduct),
      });
      
      return productId;
    } catch (error) {
      console.error('Error updating product:', error);
      throw error;
    }
  },
  
  // Delete a product
  deleteProduct: async (productId) => {
    try {
      if (!auth.currentUser) {
        throw new Error('User must be authenticated to delete products');
      }
      
      // Delete product from Firestore
      const productRef = doc(db, 'products', productId);
      await deleteDoc(productRef);
      
      // Send to backend to sync
      await fetch(`/api/products/${productId}`, {
        method: 'DELETE'
      });
      
      return true;
    } catch (error) {
      console.error('Error deleting product:', error);
      throw error;
    }
  }
};

// Cart functionality
const cartService = {
  // Get cart items
  getCart: () => {
    return JSON.parse(localStorage.getItem('cart') || '[]');
  },
  
  // Add item to cart
  addToCart: (product, quantity = 1) => {
    const cart = JSON.parse(localStorage.getItem('cart') || '[]');
    const existingItemIndex = cart.findIndex(item => item.id === product.id);
    
    if (existingItemIndex > -1) {
      // Update quantity if item is already in cart
      cart[existingItemIndex].quantity += quantity;
    } else {
      // Add new item
      cart.push({
        id: product.id,
        name: product.name,
        price: product.price,
        image_url: product.image_url,
        quantity: quantity
      });
    }
    
    localStorage.setItem('cart', JSON.stringify(cart));
    
    // Update cart UI
    updateCartUI();
    
    return cart;
  },
  
  // Update cart item quantity
  updateCartItem: (productId, quantity) => {
    const cart = JSON.parse(localStorage.getItem('cart') || '[]');
    const itemIndex = cart.findIndex(item => item.id === productId);
    
    if (itemIndex > -1) {
      if (quantity > 0) {
        cart[itemIndex].quantity = quantity;
      } else {
        // Remove item if quantity is 0 or less
        cart.splice(itemIndex, 1);
      }
      
      localStorage.setItem('cart', JSON.stringify(cart));
      
      // Update cart UI
      updateCartUI();
    }
    
    return cart;
  },
  
  // Remove item from cart
  removeFromCart: (productId) => {
    const cart = JSON.parse(localStorage.getItem('cart') || '[]');
    const updatedCart = cart.filter(item => item.id !== productId);
    
    localStorage.setItem('cart', JSON.stringify(updatedCart));
    
    // Update cart UI
    updateCartUI();
    
    return updatedCart;
  },
  
  // Clear cart
  clearCart: () => {
    localStorage.setItem('cart', JSON.stringify([]));
    
    // Update cart UI
    updateCartUI();
    
    return [];
  },
  
  // Calculate cart total
  getCartTotal: () => {
    const cart = JSON.parse(localStorage.getItem('cart') || '[]');
    return cart.reduce((total, item) => total + (item.price * item.quantity), 0);
  }
};

// Update cart UI
function updateCartUI() {
  const cart = cartService.getCart();
  const cartCount = document.getElementById('cart-count');
  const cartTotal = document.getElementById('cart-total');
  const cartItems = document.getElementById('cart-items');
  
  if (cartCount) {
    const itemCount = cart.reduce((total, item) => total + item.quantity, 0);
    cartCount.textContent = itemCount;
    
    // Show/hide based on whether cart has items
    if (itemCount > 0) {
      cartCount.style.display = 'inline-block';
    } else {
      cartCount.style.display = 'none';
    }
  }
  
  if (cartTotal) {
    cartTotal.textContent = `$${cartService.getCartTotal().toFixed(2)}`;
  }
  
  if (cartItems) {
    // Clear existing items
    cartItems.innerHTML = '';
    
    if (cart.length === 0) {
      cartItems.innerHTML = '<div class="text-center p-3">Your cart is empty</div>';
    } else {
      // Add each item to the cart UI
      cart.forEach(item => {
        const itemElement = document.createElement('div');
        itemElement.className = 'cart-item d-flex align-items-center p-2 border-bottom';
        
        itemElement.innerHTML = `
          <div class="cart-item-image me-2">
            ${item.image_url ? `<img src="${item.image_url}" alt="${item.name}" width="50" height="50" class="rounded">` : 
            '<div class="bg-secondary rounded" style="width:50px;height:50px;"></div>'}
          </div>
          <div class="cart-item-details flex-grow-1">
            <div class="cart-item-name">${item.name}</div>
            <div class="cart-item-price">$${item.price.toFixed(2)} x ${item.quantity}</div>
          </div>
          <div class="cart-item-actions">
            <button class="btn btn-sm btn-outline-danger remove-from-cart" data-product-id="${item.id}">
              <i class="fa fa-trash"></i>
            </button>
          </div>
        `;
        
        cartItems.appendChild(itemElement);
        
        // Add event listener to remove button
        const removeButton = itemElement.querySelector('.remove-from-cart');
        removeButton.addEventListener('click', () => {
          cartService.removeFromCart(item.id);
        });
      });
    }
  }
}

// Initialize Google Pay
function initGooglePay() {
  // Load Google Pay API if available
  if (window.google && window.google.payments) {
    const googlePayClient = new google.payments.api.PaymentsClient({
      environment: 'TEST' // Change to 'PRODUCTION' for real payments
    });
    
    const baseRequest = {
      apiVersion: 2,
      apiVersionMinor: 0
    };
    
    const allowedPaymentMethods = [{
      type: 'CARD',
      parameters: {
        allowedAuthMethods: ['PAN_ONLY', 'CRYPTOGRAM_3DS'],
        allowedCardNetworks: ['AMEX', 'DISCOVER', 'MASTERCARD', 'VISA']
      },
      tokenizationSpecification: {
        type: 'PAYMENT_GATEWAY',
        parameters: {
          gateway: 'example',
          gatewayMerchantId: 'exampleGatewayMerchantId'
        }
      }
    }];
    
    const isReadyToPayRequest = Object.assign({}, baseRequest);
    isReadyToPayRequest.allowedPaymentMethods = allowedPaymentMethods;
    
    // Check if Google Pay is available
    googlePayClient.isReadyToPay(isReadyToPayRequest)
      .then(response => {
        if (response.result) {
          // Google Pay is available
          const button = document.getElementById('google-pay-button');
          if (button) {
            button.style.display = 'block';
            
            // Create Google Pay button
            googlePayClient.createButton({
              onClick: () => {
                // Create payment data request
                const paymentDataRequest = Object.assign({}, baseRequest);
                paymentDataRequest.allowedPaymentMethods = allowedPaymentMethods;
                paymentDataRequest.transactionInfo = {
                  totalPriceStatus: 'FINAL',
                  totalPrice: cartService.getCartTotal().toString(),
                  currencyCode: 'USD',
                  countryCode: 'US'
                };
                paymentDataRequest.merchantInfo = {
                  merchantName: 'Women Entrepreneurs Hub'
                };
                
                // Show Google Pay payment sheet
                googlePayClient.loadPaymentData(paymentDataRequest)
                  .then(paymentData => {
                    // Process payment data
                    console.log('Payment successful:', paymentData);
                    
                    // Create order
                    createOrder(paymentData);
                  })
                  .catch(error => {
                    console.error('Error processing payment:', error);
                  });
              }
            });
          }
        }
      })
      .catch(error => {
        console.error('Error checking Google Pay availability:', error);
      });
  }
}

// Create order after successful payment
async function createOrder(paymentData) {
  try {
    if (!auth.currentUser) {
      throw new Error('User must be authenticated to place an order');
    }
    
    const cart = cartService.getCart();
    
    if (cart.length === 0) {
      throw new Error('Cannot create order with empty cart');
    }
    
    // Create order in Firestore
    const order = {
      user_id: auth.currentUser.uid,
      items: cart,
      total_amount: cartService.getCartTotal(),
      status: 'completed',
      payment_method: 'Google Pay',
      payment_id: paymentData.paymentMethodData.tokenizationData.token,
      created_at: new Date(),
      updated_at: new Date()
    };
    
    const orderRef = await addDoc(collection(db, 'orders'), order);
    
    // Send to backend to sync
    await fetch('/api/orders', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        ...order,
        firebase_id: orderRef.id
      }),
    });
    
    // Clear cart
    cartService.clearCart();
    
    // Show success message
    alert('Order placed successfully!');
    
    // Redirect to order confirmation page
    window.location.href = `/orders/${orderRef.id}`;
  } catch (error) {
    console.error('Error creating order:', error);
    alert(`Error placing order: ${error.message}`);
  }
}

// Initialize e-commerce functionality when document is loaded
document.addEventListener('DOMContentLoaded', () => {
  // Update cart UI
  updateCartUI();
  
  // Initialize Google Pay
  if (document.getElementById('google-pay-button')) {
    initGooglePay();
  }
  
  // Handle add to cart buttons
  document.querySelectorAll('.add-to-cart').forEach(button => {
    button.addEventListener('click', async (e) => {
      const productId = e.target.getAttribute('data-product-id');
      const product = await productService.getProduct(productId);
      if (product) {
        cartService.addToCart(product);
        
        // Show success message
        const toast = new bootstrap.Toast(document.getElementById('cart-toast'));
        toast.show();
      }
    });
  });
  
  // Handle product form submission
  const productForm = document.getElementById('product-form');
  if (productForm) {
    productForm.addEventListener('submit', async (e) => {
      e.preventDefault();
      
      const formData = new FormData(productForm);
      const productData = {
        name: formData.get('name'),
        description: formData.get('description'),
        price: parseFloat(formData.get('price')),
        category: formData.get('category'),
        stock: parseInt(formData.get('stock'))
      };
      
      const imageFile = formData.get('image');
      const productId = formData.get('product_id');
      
      try {
        if (productId) {
          // Update existing product
          await productService.updateProduct(productId, productData, imageFile.size > 0 ? imageFile : null);
          alert('Product updated successfully!');
        } else {
          // Add new product
          await productService.addProduct(productData, imageFile.size > 0 ? imageFile : null);
          alert('Product added successfully!');
          productForm.reset();
        }
      } catch (error) {
        alert(`Error: ${error.message}`);
      }
    });
  }
  
  // Handle proceed to checkout button
  const checkoutButton = document.getElementById('checkout-button');
  if (checkoutButton) {
    checkoutButton.addEventListener('click', () => {
      if (!auth.currentUser) {
        alert('Please sign in to checkout');
        return;
      }
      
      if (cartService.getCart().length === 0) {
        alert('Your cart is empty');
        return;
      }
      
      window.location.href = '/checkout';
    });
  }
});

// Export cart and product services
window.cartService = cartService;
window.productService = productService;

File: routes/e_commerce.py
"""
E-commerce routes for Women Entrepreneurs Hub.
Handles product listings, cart functionality, and checkout.
"""
import logging
from flask import Blueprint, render_template, session, redirect, url_for, request, jsonify
from app import db, firestore_db
from models import User, Product, Order, OrderItem
import firebase_admin

bp = Blueprint('e_commerce', __name__, url_prefix='/e-commerce')

@bp.route('/', methods=['GET'])
def index():
    """Render the e-commerce homepage with product listings."""
    # Get category filter from query parameters
    category = request.args.get('category')
    
    # Get products from database
    products = []
    
    try:
        # Query products from the database
        if category:
            products_query = Product.query.filter_by(category=category, is_available=True)
        else:
            products_query = Product.query.filter_by(is_available=True)
        
        # Add pagination
        page = request.args.get('page', 1, type=int)
        per_page = 12  # Number of products per page
        pagination = products_query.paginate(page=page, per_page=per_page, error_out=False)
        products = pagination.items
        
        return render_template(
            'e-commerce.html',
            products=products,
            category=category,
            pagination=pagination
        )
    
    except Exception as e:
        logging.error(f"Error loading e-commerce page: {e}")
        return render_template(
            'e-commerce.html',
            products=[],
            category=category,
            error="An error occurred while loading products."
        )

@bp.route('/search', methods=['GET'])
def search():
    """Search for products."""
    query = request.args.get('q', '')
    
    if not query:
        return redirect(url_for('e_commerce.index'))
    
    try:
        # Search for products by name or description
        products_query = Product.query.filter(
            (Product.name.ilike(f'%{query}%') | 
             Product.description.ilike(f'%{query}%')) &
            (Product.is_available == True)
        )
        
        # Add pagination
        page = request.args.get('page', 1, type=int)
        per_page = 12  # Number of products per page
        pagination = products_query.paginate(page=page, per_page=per_page, error_out=False)
        products = pagination.items
        
        return render_template(
            'e-commerce.html',
            products=products,
            query=query,
            pagination=pagination
        )
    
    except Exception as e:
        logging.error(f"Error searching products: {e}")
        return render_template(
            'e-commerce.html',
            products=[],
            query=query,
            error="An error occurred while searching for products."
        )

@bp.route('/product/<int:product_id>', methods=['GET'])
def view_product(product_id):
    """View a single product."""
    try:
        # Get product from database
        product = Product.query.get_or_404(product_id)
        
        # Get similar products
        similar_products = Product.query.filter_by(
            category=product.category, 
            is_available=True
        ).filter(
            Product.id != product_id
        ).limit(4).all()
        
        return render_template(
            'product.html',
            product=product,
            similar_products=similar_products
        )
    
    except Exception as e:
        logging.error(f"Error viewing product {product_id}: {e}")
        return redirect(url_for('e_commerce.index'))

@bp.route('/cart', methods=['GET'])
def view_cart():
    """View the shopping cart."""
    return render_template('cart.html')

@bp.route('/checkout', methods=['GET'])
def checkout():
    """Checkout page."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return redirect(url_for('auth.login'))
    
    return render_template('checkout.html', user_data=user_data)

@bp.route('/add-product', methods=['GET'])
def add_product_form():
    """Render the add product form."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return redirect(url_for('auth.login'))
    
    return render_template('add-product.html', user_data=user_data)

@bp.route('/my-products', methods=['GET'])
def my_products():
    """View the current user's products."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return redirect(url_for('auth.login'))
    
    try:
        # Get user's products
        user_id = user_data.get('id')
        products = Product.query.filter_by(user_id=user_id).all()
        
        return render_template(
            'my-products.html',
            products=products,
            user_data=user_data
        )
    
    except Exception as e:
        logging.error(f"Error loading user products: {e}")
        return render_template(
            'my-products.html',
            products=[],
            user_data=user_data,
            error="An error occurred while loading your products."
        )

@bp.route('/edit-product/<int:product_id>', methods=['GET'])
def edit_product(product_id):
    """Render the edit product form."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return redirect(url_for('auth.login'))
    
    try:
        # Get product from database
        product = Product.query.get_or_404(product_id)
        
        # Check if the product belongs to the current user
        if product.user_id != user_data.get('id'):
            return redirect(url_for('e_commerce.my_products'))
        
        return render_template(
            'edit-product.html',
            product=product,
            user_data=user_data
        )
    
    except Exception as e:
        logging.error(f"Error loading product for editing: {e}")
        return redirect(url_for('e_commerce.my_products'))

@bp.route('/orders', methods=['GET'])
def my_orders():
    """View the current user's orders."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return redirect(url_for('auth.login'))
    
    try:
        # Get user's orders
        user_id = user_data.get('id')
        orders = Order.query.filter_by(user_id=user_id).order_by(Order.created_at.desc()).all()
        
        return render_template(
            'orders.html',
            orders=orders,
            user_data=user_data
        )
    
    except Exception as e:
        logging.error(f"Error loading user orders: {e}")
        return render_template(
            'orders.html',
            orders=[],
            user_data=user_data,
            error="An error occurred while loading your orders."
        )

@bp.route('/orders/<order_id>', methods=['GET'])
def view_order(order_id):
    """View a single order."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return redirect(url_for('auth.login'))
    
    try:
        # If it's a Firebase ID, get the order from Firestore
        if firestore_db and not order_id.isdigit():
            order_doc = firestore_db.collection('orders').document(order_id).get()
            if not order_doc.exists:
                return redirect(url_for('e_commerce.my_orders'))
            
            order_data = order_doc.to_dict()
            
            # Check if the order belongs to the current user
            if order_data.get('user_id') != user_data.get('firebase_uid'):
                return redirect(url_for('e_commerce.my_orders'))
            
            return render_template(
                'order-details.html',
                order=order_data,
                order_id=order_id,
                user_data=user_data,
                is_firestore=True
            )
        else:
            # Get order from database
            order = Order.query.get_or_404(order_id)
            
            # Check if the order belongs to the current user
            if order.user_id != user_data.get('id'):
                return redirect(url_for('e_commerce.my_orders'))
            
            return render_template(
                'order-details.html',
                order=order,
                user_data=user_data,
                is_firestore=False
            )
    
    except Exception as e:
        logging.error(f"Error viewing order {order_id}: {e}")
        return redirect(url_for('e_commerce.my_orders'))

@bp.route('/api/products', methods=['POST'])
def create_product():
    """Create a new product."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return jsonify({'success': False, 'message': 'Not authenticated'}), 401
    
    try:
        # Get JSON data
        data = request.json
        
        # Validate required fields
        required_fields = ['name', 'price']
        for field in required_fields:
            if field not in data:
                return jsonify({'success': False, 'message': f'Missing required field: {field}'}), 400
        
        # Get user ID
        user_id = user_data.get('id')
        
        # Create new product
        product = Product(
            user_id=user_id,
            firebase_id=data.get('firebase_id'),
            name=data.get('name'),
            description=data.get('description', ''),
            price=float(data.get('price')),
            image_url=data.get('image_url'),
            category=data.get('category'),
            stock=int(data.get('stock', 0)),
            is_available=True
        )
        
        # Add to database
        db.session.add(product)
        db.session.commit()
        
        return jsonify({
            'success': True,
            'message': 'Product created successfully',
            'product_id': product.id
        })
    
    except Exception as e:
        logging.error(f"Error creating product: {e}")
        return jsonify({'success': False, 'message': str(e)}), 500

@bp.route('/api/products/<product_id>', methods=['PUT'])
def update_product(product_id):
    """Update an existing product."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return jsonify({'success': False, 'message': 'Not authenticated'}), 401
    
    try:
        # Get product from database
        product = Product.query.get(product_id)
        if not product:
            return jsonify({'success': False, 'message': 'Product not found'}), 404
        
        # Check if the product belongs to the current user
        if product.user_id != user_data.get('id'):
            return jsonify({'success': False, 'message': 'Unauthorized'}), 403
        
        # Get JSON data
        data = request.json
        
        # Update product fields
        if 'name' in data:
            product.name = data.get('name')
        
        if 'description' in data:
            product.description = data.get('description')
        
        if 'price' in data:
            product.price = float(data.get('price'))
        
        if 'image_url' in data:
            product.image_url = data.get('image_url')
        
        if 'category' in data:
            product.category = data.get('category')
        
        if 'stock' in data:
            product.stock = int(data.get('stock'))
        
        if 'is_available' in data:
            product.is_available = bool(data.get('is_available'))
        
        # Update database
        db.session.commit()
        
        return jsonify({
            'success': True,
            'message': 'Product updated successfully'
        })
    
    except Exception as e:
        logging.error(f"Error updating product {product_id}: {e}")
        return jsonify({'success': False, 'message': str(e)}), 500

@bp.route('/api/products/<product_id>', methods=['DELETE'])
def delete_product(product_id):
    """Delete a product."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return jsonify({'success': False, 'message': 'Not authenticated'}), 401
    
    try:
        # Get product from database
        product = Product.query.get(product_id)
        if not product:
            return jsonify({'success': False, 'message': 'Product not found'}), 404
        
        # Check if the product belongs to the current user
        if product.user_id != user_data.get('id'):
            return jsonify({'success': False, 'message': 'Unauthorized'}), 403
        
        # Delete from database
        db.session.delete(product)
        db.session.commit()
        
        return jsonify({
            'success': True,
            'message': 'Product deleted successfully'
        })
    
    except Exception as e:
        logging.error(f"Error deleting product {product_id}: {e}")
        return jsonify({'success': False, 'message': str(e)}), 500

@bp.route('/api/orders', methods=['POST'])
def create_order():
    """Create a new order."""
    # Check if user is logged in
    user_data = session.get('user_data')
    if not user_data:
        return jsonify({'success': False, 'message': 'Not authenticated'}), 401
    
    try:
        # Get JSON data
        data = request.json
        
        # Validate required fields
        required_fields = ['items', 'total_amount']
        for field in required_fields:
            if field not in data:
                return jsonify({'success': False, 'message': f'Missing required field: {field}'}), 400
        
        # Get user ID
        user_id = user_data.get('id')
        
        # Create new order
        order = Order(
            user_id=user_id,
            firebase_id=data.get('firebase_id'),
            total_amount=float(data.get('total_amount')),
            status=data.get('status', 'pending'),
            payment_method=data.get('payment_method'),
            payment_id=data.get('payment_id'),
            shipping_address=data.get('shipping_address')
        )
        
        # Add to database
        db.session.add(order)
        db.session.commit()
        
        # Create order items
        items = data.get('items', [])
        for item in items:
            # Get product by Firebase ID or regular ID
            product = None
            if 'id' in item:
                if isinstance(item['id'], str) and not item['id'].isdigit():
                    # This is a Firebase ID, find the product in our database
                    product = Product.query.filter_by(firebase_id=item['id']).first()
                else:
                    # This is a regular ID
                    product = Product.query.get(item['id'])
            
            if product:
                order_item = OrderItem(
                    order_id=order.id,
                    product_id=product.id,
                    quantity=int(item.get('quantity', 1)),
                    price=float(item.get('price', product.price))
                )
                db.session.add(order_item)
        
        db.session.commit()
        
        return jsonify({
            'success': True,
            'message': 'Order created successfully',
            'order_id': order.id
        })
    
    except Exception as e:
        logging.error(f"Error creating order: {e}")
        return jsonify({'success': False, 'message': str(e)}), 500
